core/win32/loader.c

```diff
@@ -253,13 +253,18 @@ os_loader_init_prologue(void)
             LOG(GLOBAL, LOG_LOADER, 2, "initial thread TEB->NlsCache=" PFX "\n",
                 pre_nls_cache);
         }
-        if (should_swap_teb_static_tls()) {
-            pre_static_tls = d_r_get_tls(STATIC_TLS_TIB_OFFSET);
-            d_r_set_tls(STATIC_TLS_TIB_OFFSET, NULL);
-            LOG(GLOBAL, LOG_LOADER, 2,
-                "initial thread TEB->ThreadLocalStoragePointer=" PFX "\n",
-                pre_static_tls);
-        }
+        /* FIX: i#6024: when kernelbase.dll has TLS or client which will lead to
+         * crash. Solution is to always save pre_static_tls but do not set
+         * ThreadLocalStoragePointer to NULL. If we set it to NULL, it will lead to
+         * crash when application uses ThreadLocalStoragePointer (eg. __declspec(thread)
+         * is used), but client or kernelbase.dll (Win11/Win22H2) or client doesn't
+         * have TLS which will fail to restore TheadLocalStoragePointer later in
+         * calls to should_swap_teb_static_tls() as tlx_next_idx is 0
+         */
+        pre_static_tls = d_r_get_tls(STATIC_TLS_TIB_OFFSET);
+        LOG(GLOBAL, LOG_LOADER, 2,
+            "initial thread TEB->ThreadLocalStoragePointer=" PFX "\n",
+            pre_static_tls);
```

Comment:
Is it possible to add a test that crashes without the fix and passes with the fix?  I guess our current test platform is https://github.com/actions/runner-images/blob/main/images/win/Windows2019-Readme.md which is Server 2019 which I think is Windows 10 1809 which is rather old: and this issue is only on 22H2.  What application and/or client is needed to trigger this?  #6024 doesn't say.  Would some of our existing tests fail if run on 22H2 or 11?

Reply:
From what I have seen. This issue for example should happen if you use ``opcode_count.dll`` and try to load ``gdi32.dll`` from instrumented program, as ``opcode_count.dll`` has TLS, and ``gdi32.dll`` also uses TLS during ``DllMain``, or if you use ``__declspec(thread)`` in main program, while using ``opcode_count.dll`` client to instrument it. (at least in my builds ``opcode_count.dll`` comes with TLS which would increment ``tls_next_idx``)

Reply:
But this won't trigger on our existing Github Actions test machines because they are too old, unfortunately.  So we'll put this in as-is since upgrading to newer Windows on those machines is a large project in its own right.